src/sandstorm/util.c++

```diff
@@ -95,6 +96,44 @@ kj::Maybe<kj::AutoCloseFd> raiiOpenAtIfExists(
   }
 }
 
+kj::Maybe<kj::AutoCloseFd> raiiOpenAtIfExistsContained(int dirfd, kj::StringPtr name, int flags, mode_t mode) {
+  auto path = kj::Path::parse(name);
+  int parentFd = dirfd;
+  int fd;
+  KJ_SYSCALL(fd = dup(dirfd));
+  kj::AutoCloseFd file(fd);
+  char path_buf[PATH_MAX+1]; // scratch buffer for file paths
+  int symlink_limit = 16; // arbitrary limit
+
+  for(int i = 0; i < path.size(); i++) {
+    const char *part = path[i].cStr();
+    KJ_SYSCALL_HANDLE_ERRORS(fd = openat(file.get(), part, flags | O_NOFOLLOW, mode)) {
+      case ENOENT:
+        return nullptr;
+      case ELOOP:
+        {
+          if(symlink_limit == 0) {
+            KJ_FAIL_SYSCALL("openat()", error);
+          }
+          symlink_limit--;
+
+          memset(&path_buf, 0, sizeof path_buf);
+          KJ_SYSCALL(readlinkat(parentFd, part, path_buf, PATH_MAX));
```

Comment:
Just noticed when re-reading this: we never actually re-assign parentFd after its initial assignment to dirfd, so this is clearly broken. I'll find some time soonish to beef up the test suite and fix things.

Reply:
Added a bunch more tests, fixed this and one or two other issues.